cpp/src/lists/copying/concatenate.cu

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/concatenate.hpp>
+#include <cudf/detail/concatenate.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <memory>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+
+namespace {
+
+/**
+ * @brief Merges the offsets child columns of multiple list columns into one.
+ *
+ * Since offsets are all relative to the start of their respective column,
+ * all offsets are shifted to account for the new starting position
+ *
+ * @param[in] columns               Vector of lists columns to concatenate
+ * @param[in] total_list_count      Total number of lists contained in the columns
+ * @param[in] stream                CUDA stream used for device memory operations
+ * and kernel launches.
+ * @param[in] mr                    Device memory resource used to allocate the
+ * returned column's device memory.
+ */
+std::unique_ptr<column> merge_offsets(std::vector<lists_column_view> const& columns,
+                                      size_type total_list_count,
+                                      cudaStream_t stream,
+                                      rmm::mr::device_memory_resource* mr)
+{
+  // outgoing offsets
+  auto merged_offsets = cudf::make_fixed_width_column(
+    data_type{INT32}, total_list_count + 1, mask_state::UNALLOCATED, stream, mr);
+  mutable_column_device_view d_merged_offsets(*merged_offsets, 0, 0);
+
+  // merge offsets
+  // TODO : this could probably be done as a single gpu operation if done as a kernel.
```

Comment:
Vague Pseudo code to maximize parallelism: (not verified, it may contain errors)

```cpp

transform_exclusive_scan(columns) --> scanned_shift 
[] c.child().size();

transform_exclusive_scan(columns) --> scanned_count 
[] c.offsets.size() - 1;

auto bin = make_transform_iterator(counting_iterator, 
[&] (i) { upper_bound(scanned_shift, i) });

transform(counting_iterator(0), counting_iterator(total_list_count),
	d_merged_offsets.begin(),
	[&] (i) { 
		colidx=bin[i]; 
		return columns[colidx].offset(i-scanned_count[colidx])+scanned_shift[colidx];
}
);
```

Reply:
do you want to give it a try in this PR itself or a new one?

Reply:
I'd prefer to do this as a seperate issue/PR.  I think it's important to get the first wave of list functionality in (upcoming gather and parquet reader PRs) before tackling every detail.

Reply:
@karthikeyann suggest filing an issue.